### ISSUE TITLE
[Idea] Intersection (and possibly union) types

### ISSUE THREAD

**Issue by zspitz** [dotnet/vblang#47]
## Intersection types

```
Dim a As String Or Number
a = 5
a = "abcd"

'saves on function overloads
Function Intersection(a As String Or Number) As String
    If TypeOf a Is String Then
        Return a
    Else 'type is number; will require some inference here
        Return a.ToString
    End If
End Function

'multitype checking -- https://github.com/dotnet/vblang/issues/23
Dim o As Object
If TypeOf o Is String Or Number Then
Else
End If

Dim flag = True
'parentheses shouldn't be required, because we can figure out that the Or is an intersection operator
'however they would add readability
If (TypeOf o Is String Or Number) Or flag Then
Else
End If

**Comment by zspitz** [comment]
Ping https://github.com/dotnet/vblang/issues/23

**Comment by zspitz** [comment]
@AdamSpeight2008 Regarding [this](https://github.com/dotnet/roslyn/issues/15458#issuecomment-262319959):

> `if (TypeOf obj Is TypeA Or TypeB ) then` could be an issue with operator precendance.
Eg Being treated as `TypeOf( obj Is (TypeA Or TypeB ) )`

Isn't `TypeOf <x> Is <Type>` the only valid syntax for `TypeOf`? What would the meaning of `TypeOf <Boolean>` be?

**Comment by AnthonyDGreen** [comment]
@zspitz we already support a limited form of intersection types in a generic method or the because the constraints are all intersected. I've thought a lot about how useful this would be other places. In fact, when we looked at adding interfaces to the Roslyn syntax tree, one thing that stopped us was the inability to specify a parameter as SytaxNode AND IXyzSyntax. This meant sacrificing all the members on SyntaxNode when using the interface type or casting constantly, so I love the idea.

The thing that concerns me the most is, ironically, verbosity and repetition in method signatures. Your proposal for type aliasing reduces it somewhat, though. I'm less optimistic about union types though. It seems rarer that I'd need them, and a bit unwieldy to use them. I guess one benefit @gafter would bring up is that they do give you a way to enforce that a Select Case/pattern match handles all cases. I know it trees us from having to define type hierarchies that depend on base classes. But do people often have code that can work with an Apple or a CarEngine?

**Comment by zspitz** [comment]
@AnthonyDGreen @gafter I've rewritten and expanded this proposal.

**Comment by zspitz** [comment]
@AnthonyDGreen 

> they do give you a way to enforce that a Select Case/pattern match handles all cases

Could you clarify this? I don't see how the two are related.

---

> But do people often have code that can work with an Apple or a CarEngine?

No. But there is often code that works on some end-type, and there are some intermediate types which need to be mapped to the end-type somehow. Currently, such a variable has to be defined as `Object`:
```vb
Function ReturnsSingleOrMultipleStrings() As Object
End Function

Dim toPrint As Object = ReturnsSingleOrMultipleStrings
If TypeOf toPrint Is String Then toPrint = {toPrint}
For Each s In toPrint
    Console.WriteLine(s)
Next
```
which means no type-safety or Intellisense:
```vb
toPrint = new Random ' no compilation error
```
However, using union types:
```vb
Function ReturnsSingleOrMultipleStrings() As IEnumerable(Of String) Or String
End Function

Dim toPrint = ReturnsSingleOrMultipleStrings

'the following is now a compilation error; Random is not compatible with either String or IEnumerable(Of String)
'toPrint = new Random 

If TypeOf toPrint Is String Then
    '"effective type" of toPrint is now String
    Console.WriteLine(toPrint.Length) 'will now compile
    toPrint = {toPrint} 'allowed by the original definition of toPrint
    '"effective type" of toPrint is now IEnumerable(Of String)
End If

'Since all branches of the If result in toPrint being an IEnumerable(Of String), the effective type of toPrint is IEnumerable(Of String)
Dim s = toPrint.FirstOrDefault ' s is typed as String
For Each s In toPrint
    Console.WriteLine(s)
Next
```
---

> a bit unwieldy to use them

I'm not quite sure what you mean. True, it's simpler to type as `Object`, but also less typesafe. For the use case of overloads which exist only for the purpose of mapping a value from one type to the next and add no other functionality, I would think multiple overloads would be more unwieldy.

**Comment by zspitz** [comment]
Pinging @KathleenDollard

**Comment by KathleenDollard** [comment]
Interesting. I'd like other folks thoughts. I'm seeing moderate improvements in keystroke/line count with a new layer of conceptual thinking. And, I don't know how we'd implement it, particularly as a language feature, unless you envision simply expanding the shorthand approach to the longer form in the examples you showed.

**Comment by Bill-McC** [comment]
I think it adds complexity to the code: you need to have in mind different types and look at all the code branching to know what type you are dealing with. And you'd have code that would be fragile as in difficult to refactor. You'd have cases where calling Foo(bar) in the same method would result in compilation failures even though bar was in scope simply because it was a different type.
In all, I don't see what it adds compared to what it costs as suitable.

**Comment by zspitz** [comment]
@KathleenDollard 

> moderate improvements in keystroke/line count with a new layer of conceptual thinking

The primary benefit is a more precise representation of the logic behind the code, when I know a particular "thing" can be either one of two types, without having to drop down to `Object`; or if I know that this "thing" inherits from / implements multiple types from different hierarchies, I can use members from both types without casting.

> a new layer of conceptual thinking

Isn't that a good thing as long as it's within the design goals of the language?

> a new layer of conceptual thinking

I think it's not as earthshattering as all that. With `Dim s As String`, I am setting down ground rules for the proper use of the object/value behind `s` -- can be concatenated with other strings to form a new string using `+`, has a `Length` property etc. Using composable types, I am setting down a slightly more flexible ground rule -- this object/value is either a `String` or a `Number`; or this object/value both inherits `MyClass` and implements `MyInterface`.

> I don't know how we'd implement it

For **intersection types**, we could use a helper `Either(Of T1, T2)` type (inspired from a suggestion in the [corresponding C# proposal](https://github.com/dotnet/csharplang/issues/399#issuecomment-305128083):

```vb
Public Structure Either(Of T1, T2)
    ReadOnly Property Item1 As T1
    ReadOnly Property Item2 As T2
    Public ReadOnly IsFirst As Boolean?
    Sub New(Item1 As T1)
        Me.Item1 = Item1
        IsFirst = True
    End Sub
    Sub New(Item2 As T2)
        Me.Item2 = Item2
        IsFirst = False
    End Sub
End Structure
```
and the compiler rewrite could look like this:

```vb
' assignment
'Dim x As String Or Integer
'If rnd.NextDouble > 0.5 Then
    'x = "abcd"
'Else
    'x = 5
'End If

Dim x As Either(Of String, Integer)
If rnd.NextDouble > 0.5 Then
    x = New Either(Of String, Integer)("abcd")
Else
    x = New Either(Of String, Integer)(5)
End If

' see PrintStrings sub above
Sub PrintStrings(toPrint As Either(Of IEnumerable(Of String), String))
    If Not toPrint.IsFirst Then
        toPrint = New Either(Of IEnumerable(Of String), String)({toPrint.Item2})
    End If

    For Each s In toPrint.Item1
        Console.WriteLine(s)
    Next
End Sub
```

**Union types** could be implemented with a similar helper type:
```vb
Public Structure Both(Of T1, T2)
    ReadOnly Property AsT1 As T1
    ReadOnly Property AsT2 As T2
    ReadOnly Property Initialized As Boolean
    Sub New(AsT1 As T1, AsT2 As T2)
        Me.AsT1 = AsT1
        Me.AsT2 = AsT2
        Initialized = True
    End Sub
End Structure
```

and given these classes:
```vb
Public Class BaseClass
    Property I As Integer
End Class
Public Interface MyInterface
    Property J As Integer
End Interface
Public Class DerivedClass
    Inherits BaseClass
    Implements MyInterface
    Property J As Integer Implements MyInterface.J
End Class
```
the compiler rewrite could look like this:
```vb
'Dim derived = New DerivedClass
'Dim y As BaseClass And MyInterface = derived
'Console.WriteLine(y.I)
'Console.WriteLine(y.J)

Dim derived = New DerivedClass
Dim y = New Both(Of BaseClass, MyInterface)(derived, derived)
Console.WriteLine(y.AsT1.I)
Console.WriteLine(y.AsT2.J)
```

**Type checking against an intersection type** would look like this:
```vb
'If TypeOf z Is String Or Number Then
If TypeOf z Is String Or TypeOf z Is Number Then
```
**Type checking against a union type** would look like this:
```vb
'If TypeOf z Is BaseClass And MyInterface Then
If TypeOf z Is BaseClass And TypeOf z Is MyInterface Then
```

**Comment by zspitz** [comment]
@Bill-McC 

> I think it adds complexity to the code

Only as a reflection of the complexity of the underlying code logic. If my code deals with some object/value which I know is either a `String` or an `Integer`, I'll have to declare it as `Object` in order to allow for both possibilities. I can then potentially assign to it something which is neither `String` nor `Integer`. If `Option Strict` is on, I will have to check at runtime and cast to the appropriate type in order to make use of either possible value; and if `Option Strict` is off, I run the risk of using members of `String` when the value is actually `Integer`, and vice versa.

If my code deals with something that is both a `BaseClass` and implements `MyInterface`, I will also have to cast to either type in order to make use of relevant members.

> you need to have in mind different types and look at all the code branching to know what type you are dealing with.

In the scenarios where intersection types add value, you have to do that anyway: e.g. in this branch, the `Object` refers to a `String`, but in that branch the `Object` refers to an `Integer`.

> you'd have code that would be fragile as in difficult to refactor

Could you elaborate on this?

> You'd have cases where calling Foo(bar) in the same method would result in compilation failures even though bar was in scope simply because it was a different type.

Is that a bad thing? The alternative is to have a runtime failure of the call to `Foo(bar)` at this point, because `bar` refers to an object/value of the wrong type.

---

All of these seem to be criticisms of the **intersection types** part of my proposal, and are not relevant to the **union types** part.

**Comment by Bill-McC** [comment]
Don't have anything to add as such.  Not sure the case of string or int is an intersection... the only thing in common is ToString.
The case of baseclass with an interface dies require casting. In days of com, there would be an IBaseClass. Thoughts around this lead me towards duck typing, and psuedo types and interfaces. That I think is worth looking at, think it may still be open under a few proposals here.
When there us commonality, worth solving. When there is none, Object and casting is clearer .

Regards, Bill.

________________________________

**Comment by KathleenDollard** [comment]
This is the sort of thing that would want to have serious discussions about whether to do in the language (via erasure, your suggestion) or with BCL support. Either way, this is likely to be something we'd want a broad base of support and probably having C# take the lead on a rather technical issue.

**Comment by zspitz** [comment]
@KathleenDollard There is one possible reason why this would be more attractive in VB.NET over C# -- "classic" VB didn't have overloads and was limited to a single function name within a given scope; whereas C#'s syntax is in the main inspired by C-like languages, which support function overloads. I am not sure if this is still a design goal in VB.NET, but after all VB.NET had the `Optional` keyword from day one, while it took a few releases for similar functionality to be made available to C#.

Typescript can serve as a model for the transition from a language without composite types to a language with composite types; but I think the more appropriate comparison is to F#'s composite types, as F# is a statically typed language while still having composite types.

---

I'm currently rewriting the Typescript definitions for the Excel object model, and this benefit is actually quite striking. Consider this definition, autogenerated from the registered type lib information:

    interface Sheets {
        // ....
        Item(Index: any): any;
        // ...
    }

What arguments can be passed into the `Item` method? What will it return? The answer to that requires a context switch to the browser, browsing to the Excel Sheets object, finding the Item method, and reading the documentation; followed by another cognitive context switch, and hopefully passing in the arguments correctly.

Arguably, it's possible (and advisable) to embed the documentation in the file, and it would then be available in the editor. However, I still have no guarantee that the values I pass in are of the correct type for this context.

But the following definition describes the possible values simply and concisely, and also enforces it at compile time:

    interface Sheets {
        // ....
        Item(Index: string | number): Sheet;
        Item(Indexes: SafeArray<string | number>): Sheets;
        // ...
    }

Now, this could be expressed using overloads:

    interface Sheets {
        // ....
        Item(Index: string): Sheet;
        Item(Index: number): Sheet;
        Item(Indexes: SafeArray<string>): Sheets;
        Item(Indexes: SafeArray<number>): Sheets;
        // ...
    }

but now I have four method signatures to read instead of two (aside from the overloads not being quite accurate, because a single `SafeArray` could hold both numbers and strings).

**Comment by bandleader** [comment]
> @zspitz Any further thoughts on this?

Just that for me, the main use case would be union types, for simplifying overloads. 
i.e. Sometimes I can take a URI or a string, etc., because I'll just convert one to the other... so typing it as `URI Or String` would save an overload
And if I have _two_ parameters that can be either a URI or a string, then I save _three_ overloads.
And I can even abstract away the conversion code: ```vb
Function Foo(one As (URI Or String), two As (URI Or String))
  Dim getUri = Function(s As (URI Or String)) If(TypeOf s Is URI, s, New URI(s))
  Dim uriOne = getUri(one), uriTwo = getUri(two)
  'Use uriOne and uriTwo here...
End Function
```

**Comment by Berrysoft** [comment]
If you write `Either(Of T1, T2)` with two `Operator CType`:
``` VB
Public Structure Either(Of T1, T2)
    Public ReadOnly Property Item1 As T1
    Public ReadOnly Property Item2 As T2
    Public ReadOnly IsFirst As Boolean?
    Public Sub New(Item1 As T1)
        Me.Item1 = Item1
        IsFirst = True
    End Sub
    Public Sub New(Item2 As T2)
        Me.Item2 = Item2
        IsFirst = False
    End Sub
    Public Shared Widening Operator CType(value As T1) As Either(Of T1, T2)
        Return New Either(Of T1, T2)(value)
    End Operator
    Public Shared Widening Operator CType(value As T2) As Either(Of T1, T2)
        Return New Either(Of T1, T2)(value)
    End Operator
End Structure
``` 
You can write a function like this:
``` VB
Function Foo(one As Either(Of Uri, String), two As Either(Of Uri, String))
    Dim getUri = Function(s As Either(Of Uri, String)) If(s.IsFirst, s.Item1, New Uri(s.Item2))
    Dim uriOne = getUri(one), uriTwo = getUri(two)
    'Use uriOne and uriTwo here...
End Function
```
And call it like:
``` VB
Foo("https://www.google.com/", New Uri("https://github.com/"))
```

**Comment by zspitz** [comment]
Shoutout to the [OneOf](https://github.com/mcintyre321/OneOf) library, which allows creating union types in .NET.

**Comment by ris-work** [comment]
TL;DR: Now that C# is seriously moving towards Union types by **C# 16** (.NET 11), please reconsider the status about "No plans" status.

Relevant: https://github.com/dotnet/csharplang/tree/main/meetings/working-groups/discriminated-unions moved from "proposals" to "working-groups", and to ensure **feature parity**, it becomes something we should have at least partially in VB .NET (at least to the extent of being able to consume it when an **underlying C# library exports it, or vice versa**). I do know that VB is effectively "frozen" now, but the lack of unions does feel like it is frozen at an incomplete stage. It is not a "fancy" feature and other languages have had it a long time ago, like Rust, and I do use them a lot.

It is also very useful when we have `AnyOf` in JSON. It makes things easier when you can just define a union of potential responses instead of first casting it to an `object` and then try casting it again... and again... or using third-party dependencies for what I believe should be supported by the language/runtime.

Relevant issue(s) in other CLR language(s):
https://github.com/dotnet/csharplang/issues/8928

Moreover, F# has this feature already (since... long ago), though it is implemented differently (another comment has already mentioned it).
